File: shoring/compat/src/main/java/com/panopset/compat/Fileop.java
package com.panopset.compat;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.StringWriter;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

public class Fileop {

  private Fileop() {}

  private static final String NULL_FILE_MSG = "file is null";

  public static File getUserHome() {
    return new File(System.getProperty("user.home"));
  }

  public static void touch(File file) throws IOException {
    if (prepFileForWriting(file) && !file.createNewFile()) {
      Logop.warn(String.format("Unable to create %s", Fileop.getCanonicalPath(file)));
    }
  }

  public static void deleteLines(File file, final String searchCriteria) {
    if (file == null) {
      Logop.error(NULL_FILE_MSG);
      return;
    }
    Logop.green(String.format("Found file: %s", Fileop.getCanonicalPath(file)));
    if (file.isDirectory()) {
      DirectoryProcessor.Listener listener = new DirectoryProcessor.Listener() {

        @Override
        public boolean processFile(File file) {
          return deleteLinesInFile(file, searchCriteria);
        }
      };
      new DirectoryProcessor(file, listener).exec();
    } else {
      deleteLinesInFile(file, searchCriteria);
      Logop.green(
          String.format(Stringop.CS, Nls.xlate("Processed file"), Fileop.getCanonicalPath(file)));
    }
  }

  private static boolean deleteLinesInFile(File file, String searchCriteria) {
    if (searchCriteria == null) {
      return false;
    }
    final List<String> targets = Stringop.stringToList(searchCriteria);
    if (targets.isEmpty()) {
      return false;
    }
    new FileProcessor(file).withLineFilter(new LineFilter() {
      @Override
      public String filter(String str) {
        for (String s : targets) {
          if (str.contains(s)) {
            return null;
          }
        }
        return str;
      }
    }).exec();
    return true;
  }

  public static String getStandardPath(File file) {
    return getCanonicalPath(file).replace("\\", "/");
  }

  public static String getParentDirectory(final File file) {
    if (file.exists()) {
      File fullFile = new File(getCanonicalPath(file));
      return getCanonicalPath(fullFile.getParentFile());
    }
    return getCanonicalPath(file.getParentFile());
  }

  public static boolean fileExists(File file) {
    if (file == null) {
      return false;
    } else {
      return file.exists() && file.isFile() && file.canRead();
    }
  }
  public static boolean dirExists(File dir) {
    if (dir == null) {
      return false;
    } else {
      return dir.exists() && dir.isDirectory() && dir.canRead();
    }
  }

  public static File createTempFile(String fileName) {
    return new File(combinePaths(Stringop.TEMP_DIR_PATH, fileName));
  }

  public static void write(String[] strs, File file) {
    if (!prepFileForWriting(file)) {
      return;
    }
    try (FileWriter fw = new FileWriter(file);

        BufferedWriter bw = new BufferedWriter(fw)) {
      for (String s : strs) {
        bw.write(s);
        bw.write(Stringop.EOL);
      }
    } catch(IOException ex) {
      Logop.error(ex);
    }
  }

  public static void write(List<String> strs, File file) {
    if (!prepFileForWriting(file)) {
      return;
    }
    try (FileWriter fw = new FileWriter(file);

        BufferedWriter bw = new BufferedWriter(fw)) {
      for (String s : strs) {
        bw.write(s);
        bw.write(Stringop.EOL);
      }
    } catch(IOException ex) {
      Logop.error(ex);
    }
  }

  public static void write(String str, File file) {
    if (!prepFileForWriting(file)) {
      return;
    }
    try (FileWriter fw = new FileWriter(file);

        BufferedWriter bw = new BufferedWriter(fw)) {
      bw.write(str);
    } catch (IOException ex) {
      Logop.error(ex);
    }
  }

  private static boolean prepFileForWriting(File file) {
    if (file == null) {
      Logop.error("Can't write to a null file");
      return false;
    }
    if (Stringop.isBlank(file.getPath())) {
      Logop.error("Can't write to blank file path.");
      return false;
    }
    File parentFile = file.getParentFile();
    if (parentFile == null) {
      Logop.error("Unable to write to file with no parent directory.");
      return false;
    }
    file.getParentFile().mkdirs();
    return true;
  }

  public static String getCanonicalPath(File file) {
    if (file == null) {
      return "";
    }
    if (Stringop.isBlank(file.getPath())) {
      return "";
    }
    try {
      return file.getCanonicalPath();
    } catch (IOException ex) {
      Logop.error(ex);
      return ex.getMessage();
    }
  }

  public static boolean isFileOneOfExtensions(File file, String extsToTry) {
    if (extsToTry != null && file != null) {
      String fileExtension = getExtension(file);
      if (extsToTry.contains(",")) {
        String[] exts = extsToTry.split(",");
        for (String ext : exts) {
          if (ext.trim().equalsIgnoreCase(fileExtension)) {
            return true;
          }
        }
        return false;
      } else {
        return getExtension(file).equalsIgnoreCase(extsToTry);
      }
    } else {
      return false;
    }
  }

  public static String getExtension(String fileName) {
    int i = fileName.lastIndexOf(46);
    return i > 0 ? fileName.substring(i + 1) : "";
  }

  public static String getExtension(File file) {
    return getExtension(file.getName());
  }

  public static String combinePaths(String dirPath, String path) {
    return combinePaths(new File(dirPath), path);
  }

  public static String combinePaths(File dir, String path) {
    return String.format("%s%s%s", getCanonicalPath(dir), FILE_SEP, path);
  }

  public static String readTextFile(String filePath) throws IOException {
    return readTextFile(new File(filePath));
  }

  public static String readTextFile(File file) {
    if (!file.exists()) {
      Logop.warn(String.format("File %s doesn't exist.", getCanonicalPath(file)));
      return "";
    }
    StringWriter sw = new StringWriter();
    try {
      Streamop.copyChars(new FileReader(file), sw);
    } catch (IOException ex) {
      Logop.error(file, ex);
    }
    return sw.toString();
  }

  public static List<String> readLines(File file) {
    if (!file.exists()) {
      Logop.error("File does not exist", file);
      return new ArrayList<>();
    }
    try {
      return Streamop.getLinesFromReader(new FileReader(file));
    } catch (FileNotFoundException ex) {
      Logop.error(ex);
      return new ArrayList<>();
    }
  }

  public static void delete(final File file) {
    if (file == null) {
      return;
    }
    if (file.isDirectory()) {
      File[] files = file.listFiles();
      if (files != null) {
        for (File f : files) {
          delete(f);
        }
      }
    }
    if (file.exists()) {
      try {
        Files.delete(file.toPath());
      } catch (IOException e) {
        Logop.error(e);
      }
    }
  }

  public static void mkdirs(final File path) {
    if (!path.mkdirs()) {
      Logop.error(
          String.format("%s %s", Nls.xlate("Unable to create path to"), getCanonicalPath(path)));
    }
  }

  public static void copyInputStreamToFile(final InputStream inp, String fileName)
      throws IOException {
    copyInputStreamToFile(inp, new File(fileName));
  }

  public static void copyInputStreamToFile(final InputStream is, File file) throws IOException {
    Streamop.streamToWriter(is, new FileWriter(file));
  }

  public static void moveFile(File ff, File tf) throws IOException {
    Files.move(ff.toPath(), tf.toPath());
  }

  public static void copyFile(File ff, File tf) throws IOException {
    Files.copy(ff.toPath(), tf.toPath());
  }

  public static final File TEMP_DIRECTORY = new File(Stringop.USER_HOME + "/temp");
  public static final String CRLF = "\r\n";
  public static final String FILE_SEP = System.getProperty("file.separator");
  public static final String PATH_SEP = System.getProperty("path.separator");

  public static String removeExtension(String str) {
    int i = str.lastIndexOf(".");
    if (i > -1) {
      return str.substring(0, i);
    }
    return str;
  }

  public static boolean checkParent(File targetFile) {
    File parent = targetFile.getParentFile();
    if (parent.exists()) {
      return true;
    } else {
      return parent.mkdirs();
    }
  }
}

File: code/shoring/compat/src/main/kotlin/com/panopset/compat/Fileop.kt
package com.panopset.compat

import com.panopset.compat.Streamop.copyChars
import com.panopset.compat.Streamop.getLinesFromReader
import com.panopset.compat.Streamop.streamToWriter
import com.panopset.compat.Stringop.FSP
import com.panopset.compat.Stringop.getEol
import com.panopset.compat.Stringop.isBlank
import java.io.*
import java.nio.file.Files
import java.util.*
import kotlin.collections.ArrayList

object Fileop {

    val userHome = File(System.getProperty("user.home"))

    fun touch(file: File) {
        if (prepFileForWriting( file) && !file.createNewFile()) {
            Logz.warn("Unable to create ${getCanonicalPath(file)}")
        }
    }

    fun getStandardPath(file: File): String {
        return getCanonicalPath(file).replace("\\", "/")
    }

    fun getParentDirectory(file: File): String {
        if (file.exists()) {
            val fullFile = File(getCanonicalPath(file))
            return getCanonicalPath(fullFile.parentFile)
        }
        return getCanonicalPath(file.parentFile)
    }

    fun fileExists(file: File?): Boolean {
        return if (file == null) {
            false
        } else {
            file.exists() && file.isFile && file.canRead()
        }
    }

    fun dirExists(dir: File): Boolean {
        return dir.exists() && dir.isDirectory && dir.canRead()
    }

    fun createTempFile(fileName: String): File {
        return File(combinePaths(Stringop.TEMP_DIR_PATH, fileName))
    }

    fun write(strs: Array<String>, file: File) {
        if (!prepFileForWriting( file)) {
            return
        }
        try {
            FileWriter(file).use { fw ->
                BufferedWriter(fw).use { bw ->
                    for (s in strs) {
                        bw.write(s)
                        bw.write(getEol())
                    }
                }
            }
        } catch (ex: IOException) {
            Logz.errorEx(ex)
        }
    }

    fun write(strs: List<String>, file: File) {
        if (!prepFileForWriting( file)) {
            return
        }
        try {
            FileWriter(file).use { fw ->
                BufferedWriter(fw).use { bw ->
                    for (s in strs) {
                        bw.write(s)
                        bw.write(getEol())
                    }
                }
            }
        } catch (ex: IOException) {
            Logz.errorEx(ex)
        }
    }

	fun write(str: String, file: File) {
        if (!prepFileForWriting( file)) {
            return
        }
        try {
            FileWriter(file).use { fw -> BufferedWriter(fw).use { bw -> bw.write(str) } }
        } catch (ex: IOException) {
            Logz.errorEx(ex)
        }
    }

    private fun prepFileForWriting(file: File): Boolean {
        if (isBlank(file.path)) {
            Logz.errorMsg("Can't write to blank file path.")
            return false
        }
        val parentFile = file.parentFile
        if (parentFile == null) {
            Logz.errorMsg("Unable to write to file with no parent directory.")
            return false
        }
        file.parentFile.mkdirs()
        return true
    }

	fun getCanonicalPath(file: File?): String {
        if (file == null) {
            return ""
        }
        return file.canonicalPath
    }

    fun isFileOneOfExtensions(file: File, extsToTry: String): Boolean {
            val fileExtension = getExtension(file)
            if (extsToTry.contains(",")) {
                val exts = extsToTry.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                for (ext in exts) {
                    if (ext.trim { it <= ' ' }.equals(fileExtension, ignoreCase = true)) {
                        return true
                    }
                }
                return false
            } else {
                return getExtension(file).equals(extsToTry, ignoreCase = true)
            }
    }

	fun getExtension(fileName: String): String {
        val i = fileName.lastIndexOf(46.toChar())
        return if (i > 0) fileName.substring(i + 1) else ""
    }

    fun getExtension(file: File): String {
        return getExtension(file.getName())
    }

    fun combinePaths(dirPath: String, path: String): String {
        return combinePaths(File(dirPath), path)
    }

    fun combinePaths(dir: File, path: String): String {
        return String.format("%s%s%s", getCanonicalPath(dir), FSP, path)
    }

    fun readTextFile(filePath: String): String {
        return readTextFile( File(filePath))
    }

    fun readTextFile(file: File): String {
        if (!file.exists()) {
            Logz.warn(String.format("File %s doesn't exist.", getCanonicalPath(file)))
            return ""
        }
        val sw = StringWriter()
        try {
            copyChars(FileReader(file), sw)
        } catch (ex: IOException) {
            Logz.errorMsg(file, ex)
        }
        return sw.toString()
    }

	fun readLines(file: File): List<String> {
        if (!file.exists()) {
            Logz.errorMsg("File does not exist", file)
            return ArrayList()
        }
        return try {
            getLinesFromReader(FileReader(file))
        } catch (ex: FileNotFoundException) {
            Logz.errorEx(ex)
            ArrayList()
        }
    }

    fun delete(file: File) {
        if (file.isDirectory) {
            val files = file.listFiles()
            if (files != null) {
                for (f in files) {
                    delete( f)
                }
            }
        }
        if (file.exists()) {
            try {
                Files.delete(file.toPath())
            } catch (e: IOException) {
                Logz.errorEx(e)
            }
        }
    }

    fun mkdirs(path: File) {
        if (!path.mkdirs()) {
            Logz.errorMsg("${Nls.xlate("Unable to create path to")},${getCanonicalPath(path)}")
        }
    }

    fun copyInputStreamToFile(inp: InputStream, fileName: String) {
        copyInputStreamToFile(inp, File(fileName))
    }

    fun copyInputStreamToFile(inputStream: InputStream, file: File) {
        streamToWriter(inputStream, FileWriter(file))
    }

    fun moveFile(ff: File, tf: File) {
        Files.move(ff.toPath(), tf.toPath())
    }

    fun copyFile(ff: File, tf: File) {
        try {
            Files.copy(ff.toPath(), tf.toPath())
        } catch (e: IOException) {
            Logz.errorEx(e)
        }
    }

    val TEMP_DIRECTORY = File(Stringop.USH + "/Documents/temp")

    fun removeExtension(str: String): String {
        val i = str.lastIndexOf(".")
        return if (i > -1) {
            str.substring(0, i)
        } else str
    }

	fun checkParent(targetFile: File): Boolean {
        val parent = targetFile.parentFile
        return if (parent.exists()) {
            true
        } else {
            parent.mkdirs()
        }
    }

    fun loadProps(fileName: String): Properties {
        return loadProps(File(fileName))
    }

    fun loadProps(file: File): Properties {
        val rtn = Properties()
        try {
            FileReader(file).use { fileReader ->
                BufferedReader(fileReader).use {
                    rtn.load(it)
                }
            }
        } catch (ex: Exception) {
            Logz.errorMsg(file, ex)
        }
        return rtn
    }
}

fun props2map(props: Properties): MutableMap<String, String> {
    val rtn = HashMap<String, String>()
    for (e in props.entries) {
        rtn[e.key.toString()] = e.value.toString()
    }
    return rtn
}
